keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+kind: PolicyBinding
+metadata:
+  name: "xyzlimit-scale.example.com"
+spec:
+  policy: "replicalimit-policy.example.com"
+  params: "replica-limit-nontest.example.com"
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: Exists
+  enforcement: [Deny]
+```
+
+With this binding, the test and global policy bindings overlap. Resources
+admitted to test environment would then be checked against both policy
+configurations.
+
+To summarize, this "API Shape" separates admission policy _definition_ from
+_configuration_. This has a couple advantages:
+
+- Access to, and delegation of, policy configuration is more manageable. In
+  particular, Kubernetes RBAC works well with this design.
+- Without the separation, the next most obvious API shape would be to encode
+  everything into a single resource, which could easily become very large and
+  run into resource size limits.
+
+##### Policy Definitions
+
+Policy definitions are responsible for:
+
+- Defining what validations the policy enforces and how violations are reported 
+- Defining how a policy may be configured
+
+Each `ValidatingAdmissionPolicy` resource contains a `spec.match` to declare
+what resources it validates. This field is required.
+
+- `spec.match` constrains which resources this policy can be applied to. Policy
+  bindings each have match rules with further narrow this constraint, but cannot
+  expand it. This allows the CEL expressions to make safe assumptions. E.g. a
+  CEL expression that is constrained to CREATES and UPDATES of resources is
+  guaranteed the root `object` variable is never null, but a CEL expression that
+  might need to evaluate a DELETE must handle the root `object` variable being
+  null. See below "Match criteria" section for how match criteria is described.
+- `spec.match` guides CEL expression type checking, see the below "Type safety"
+  section for more details. 
+
+CEL expressions have access to the contents of the `AdmissionReview` type,
+organized into CEL variables as well as some other useful variables:
+
+- 'object'
+- 'oldObject'
+- 'review'
+  - 'requestResource' (GVR)
+  - 'resource' (GVR)
+  - 'name'
+  - 'namespace'
+  - 'operation'
+  - 'userInfo'
+  - 'dryRun'
+  - 'options'
+- 'config' - configuration data of the policy configuration being validated
+
+See below "Reporting violations to Clients" for more detail about how the
+`spec.validations` field works and how violations are reported.
+
+##### Policy Configuration
+
+`PolicyBinding` resources and parameter CRDs together define how cluster
+administrators configure policies for clusters.
+
+Each `PolicyBinding` contains:
+
+- `spec.policy` - A reference to the policy being configured
+- `spec.match` - Match criteria for which resources the policy should validate
+- `spec.params` - Reference to the custom resource containing the params to use
+  when validating resources 
+- `spec.enforcement` - How validation failures should be enforced
```

Comment:
I think enforcement should have a narrower scope as it looks like it's combining many different policy types and concepts in one here. 

Consider that the admission policy effect is always to reject the request, but the error code determines whether the reason for the rejection is user-visible or not. Should this decision about the error code belong with the person who configure the policy or the person who configures the binding? My bias is toward the person who configures the policy. 

Now, let's say the person creating the policy binding wants to control rollout of the binding, shifting it from a dry-run state which audit logs rejections or possibly audit logs differences between the itself and another binding. That's a very reasonable decision to put into the hands of the person who creates the binding. However, I wouldn't call this `enforcement`, but instead `rollout`. If left unspecified the `rollout` would be `live`.

Ideally, the decision to reject the request is separate from the decision to deny (or grant) the request. Until such. a policy exists, I would consider always auditing permission denied responses (as opposed to invalid argument responses).

Reply:
I agree that either params or binding is likely a better place for this. But we have to define what happens when multiple params are bound, or one param applies via multiple bindings.

Reply:
It might be worth reconsidering a more complex syntax for `enforcement` to allow a bit more decoupling between "signal sent by policy" and "action taken by enforcement point"